## Gather

In [126]:
import pandas as pd
import numpy as np

In [127]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [128]:
patients.head(20)

patient_id assigned_sex  given_name     surname  \
0            1       female         Zoe     Wellish   
1            2       female      Pamela        Hill   
2            3         male         Jae      Debord   
3            4         male        Liêm        Phan   
4            5         male         Tim     Neudorf   
5            6         male      Rafael       Costa   
6            7       female        Mary       Adams   
7            8       female      Xiuxiu       Chang   
8            9         male       Dsvid  Gustafsson   
9           10       female      Sophie     Cabrera   
10          11       female       Sandy  Gunnarsson   
11          12         male   Abdul-Nur         Isa   
12          13         male  Omeokachie    Ibeamaka   
13          14       female    Anenechi       Chidi   
14          15       female        Asia     Woźniak   
15          16         male       Søren        Lund   
16          17       female         Tám        Liễu   
17          18       female     Roxanne   Andreyeva   
18          19         male     William       Oates   
19          20         male         Zak       Kelly   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
5           1140 Willis Avenue      Daytona Beach     Florida   32114.0   
6             3145 Sheila Lane            Burbank          NV   84728.0   
7   2687 Black Oak Hollow Road        Morgan Hill          CA   95037.0   
8           1790 Nutter Street        Kansas City          MO   64105.0   
9            3303 Anmoore Road           New York    New York   10011.0   
10          87 Wood Duck Drive            Rudyard          MI   49780.0   
11      1092 Farm Meadow Drive          Brentwood          TN   37027.0   
12          2544 Worley Avenue          Lynchburg          VA   24504.0   
13            826 Broad Street         Birmingham          AL   35203.0   
14       4970 Heather Sees Way              Tulsa          OK   74105.0   
15      2438 Shady Pines Drive          Kingsport          VA   37660.0   
16          2152 Heritage Road             Fresno  California   93706.0   
17         2103 Edington Drive             Smyrna          GA   30082.0   
18            441 Tibbs Avenue            Ekalaka          MT   59324.0   
19    994 Hill Croft Farm Road           Oroville  California   95966.0   

          country                                        contact   birthdate  \
0   United States           951-719-9170ZoeWellish@superrito.com   7/10/1976   
1   United States          PamelaSHill@cuvox.de+1 (217) 569-3204    4/3/1967   
2   United States               402-363-6804JaeMDebord@gustr.com   2/19/1980   
3   United States     PhanBaLiem@jourrapide.com+1 (732) 636-8246   7/26/1951   
4   United States                334-515-7487TimNeudorf@cuvox.de   2/18/1928   
5   United States       386-334-5237RafaelCardosoCosta@gustr.com   8/31/1931   
6   United States              775-533-5933MaryBAdams@einrot.com  11/19/1969   
7   United States           XiuxiuChang@einrot.com1 408 778 3236   8/13/1958   
8   United States        816-265-9578DavidGustafsson@armyspy.com    3/6/1937   
9   United States  SophieCabreraIbarra@teleworm.us1 718 795 9124   12/3/1930   
10  United States         906-478-8949SandyGunnarsson@dayrep.com   7/16/1974   
11  United States     Abdul-NurMummarIsa@rhyta.com1 931 207 0839    2/3/1954   
12  United States      OmeokachieIbeamaka@einrot.com434-509-2614    8/5/1957   
13  United States     AnenechiChidi@armyspy.com+1 (205) 417-8095    3/7/1961   
14  United States              AsiaWozniak@rhyta.com918-712-3469   8/15/

In [130]:
treatments.head(20)

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
5      jasmine         sykes          -  42u - 44u         7.56       7.18   
6       sophia        haugen  37u - 42u          -         7.65       7.27   
7        eddie        archer  31u - 38u          -         7.89       7.55   
8        saber        ménard          -  54u - 54u         8.08       7.70   
9         asia       woźniak  30u - 36u          -         7.76       7.37   
10      joseph           day  29u - 36u          -         7.70       7.19   
11   kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12     roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13      gregor          bole          -  47u - 45u         7.61       7.16   
14      simone  baumgaertner  27u - 37u          -         7.74       7.30   
15        enco        žibrik  55u - 68u          -         7.78       7.34   
16     camilla      zaitseva  28u - 37u          -         7.53       7.13   
17        gina          cain          -  36u - 36u         7.88       7.40   
18  addolorata      lombardi          -  49u - 46u         7.75       7.33   
19      khalid      johnsrud          -  54u - 54u         8.35       7.94   

    hba1c_change  
0            NaN  
1           0.97  
2            NaN  
3           0.35  
4           0.32  
5           0.38  
6           0.38  
7           0.34  
8            NaN  
9            NaN  
10           NaN  
11          0.38  
12           NaN  
13          0.95  
14           NaN  
15           NaN  
16           NaN  
17          0.98  
18           NaN  
19           NaN

In [132]:
adverse_reactions.head(20)

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache

In [133]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [134]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [135]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [136]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [137]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [138]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [139]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [140]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [141]:
patients.sample(5)

patient_id assigned_sex given_name     surname                  address  \
154         155       female       Maya    Isaksson       1333 Comfort Court   
153         154         male       John    Carreiro  1463 Martha Ellen Drive   
488         489         male       Mika  Martinsson        962 George Street   
342         343       female   Mathilde    Nørgaard      846 Copperhead Road   
50           51         male    Stephen      Gibson   283 Simons Hollow Road   

             city    state  zip_code        country  \
154       Madison       WI   53711.0  United States   
153          Reno       NV   89509.0  United States   
488         Ocala  Florida   34471.0  United States   
342      Hartford       CT    6182.0  United States   
50   Wilkes Barre       PA   18701.0  United States   

                                          contact  birthdate  weight  height  \
154     MayaIsaksson@jourrapide.com1 608 277 0146  7/14/1959   137.7      67   
153     JohnACarreiro@superrito.com1 775 770 7827   4/8/1976   177.8      68   
488        352-453-4601MikaMartinsson@armyspy.com  1/27/1970   165.0      67   
342  MathildeDNrgaard@superrito.com1 860 618 4560  1/21/1985   127.2      61   
50             StephenGibson@cuvox.de570-819-1140  4/25/1991   238.9      70   

      bmi  
154  21.6  
153  27.0  
488  25.8  
342  24.0  
50   34.3

In [142]:
patients.surname.value_counts()

Doe             6
Jakobsen        3
Taylor          3
Johnson         2
Cabrera         2
Tạ              2
Grímsdóttir     2
Bùi             2
Gersten         2
Hueber          2
Silva           2
Kadyrov         2
Liễu            2
Lương           2
Ogochukwu       2
Dratchev        2
Lund            2
Souza           2
Aranda          2
Tucker          2
Berg            2
Schiavone       2
Collins         2
Nilsen          2
Kowalczyk       2
Batukayev       2
Correia         2
Parker          2
Woźniak         2
Lâm             2
               ..
Tomaszewski     1
Lukić           1
Montez          1
Pospíšil        1
Chandler        1
Nebay           1
Fodor           1
Csonka          1
Azevedo         1
Bole            1
Okoli           1
Crawford        1
Kristiansen     1
Ferreira        1
Grant           1
Brhane          1
Baum            1
Achebe          1
O'Brian         1
Onyekaozulu     1
Sleiman         1
Ekechukwu       1
Tobeolisa       1
Koldenhof       1
Guðbrandss

In [143]:
patients.address.value_counts()

123 Main Street             6
2476 Fulton Street          2
2778 North Avenue           2
648 Old Dear Lane           2
1094 Jones Avenue           1
1324 Duffy Street           1
2808 Circle Drive           1
2166 Tipple Road            1
4982 Wood Street            1
1763 Williams Mine Road     1
631 Isaacs Creek Road       1
163 Hide A Way Road         1
4851 Andy Street            1
4970 Heather Sees Way       1
4435 Poe Road               1
200 Hall Place              1
2881 Harter Street          1
3683 Gorby Lane             1
192 Patton Lane             1
57 Norman Street            1
846 Copperhead Road         1
4890 Abia Martin Drive      1
3077 Colony Street          1
2784 Turkey Pen Road        1
522 Lamberts Branch Road    1
4500 Myra Street            1
3261 Desert Broom Court     1
288 Colony Street           1
3259 Roy Alley              1
541 Green Hill Road         1
                           ..
3006 Maple Court            1
230 Lady Bug Drive          1
3094 Oral 

In [144]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [145]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [146]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [147]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [148]:
sum(treatments.auralin.isnull())

0

In [149]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean yet)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [150]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

#### `treatments`: Missing records (280 instead of 350)

##### Define
Import the cut treatments into a DataFrame and concatenate it with the original treatments DataFrame.

##### Code

In [151]:
treatments_cut = pd.read_csv('treatments_cut.csv')
treatments_clean = pd.concat([treatments_clean, treatments_cut],
                             ignore_index=True)

##### Test

In [152]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [153]:
treatments_clean.tail()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change  
345          0.34  
346          0.95  
347          0.30  
348           NaN  
349           NaN

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)

##### Define
Recalculate the `hba1c_change` column: `hba1c_start` minus `hba1c_end`. 

##### Code

In [154]:
treatments_clean.hba1c_change = (treatments_clean.hba1c_start - 
                                 treatments_clean.hba1c_end)

##### Test

In [155]:
treatments_clean.hba1c_change.head()

0    0.43
1    0.47
2    0.43
3    0.35
4    0.32
Name: hba1c_change, dtype: float64

### Tidiness

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
Extract the *phone number* and *email* variables from the *contact* column using regular expressions and pandas' `str.extract` method. Drop the *contact* column when done.

##### Code

In [156]:
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.][a-zA-Z]+)', expand=True)
# Note: axis=1 denotes that we are referring to a column, not a row
patients_clean = patients_clean.drop('contact', axis=1)

##### Test

In [157]:
# Confirm contact column is gone
list(patients_clean)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'birthdate',
 'weight',
 'height',
 'bmi',
 'phone_number',
 'email']

In [158]:
patients_clean.phone_number.sample(25)

391         786-970-4206
426         305-434-6299
90     +1 (585) 902-9127
451         909 982 4264
17          678-829-8578
70          740 629 5172
500         816-223-6007
5           386-334-5237
349         360-494-5489
330    +1 (843) 368-5129
409         716-897-5050
264                  NaN
44          207 861 4587
53          617-317-5055
198    +1 (859) 977-9462
351         423-799-1730
126         508-739-5632
492         714-496-2264
124    +1 (508) 526-3432
417         360-689-4989
241         901-484-5225
149         830-640-5848
270    +1 (618) 512-3319
482         361-693-4960
57          858-413-3163
Name: phone_number, dtype: object

In [159]:
# Confirm that no emails start with an integer (regex didn't match for this)
patients_clean.email.sort_values().head()

404               AaliyahRice@dayrep.com
11          Abdul-NurMummarIsa@rhyta.com
332                AbelEfrem@fleckens.hu
258              AbelYonatan@teleworm.us
305    AddolorataLombardi@jourrapide.com
Name: email, dtype: object

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
Melt the *auralin* and *novodra* columns to a *treatment* and a *dose* column (dose will still contain both start and end dose at this point). Then split the dose column on ' - ' to obtain *start_dose* and *end_dose* columns. Drop the intermediate *dose* column.

##### Code

In [160]:
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')
treatments_clean = treatments_clean[treatments_clean.dose != "-"]
treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str
treatments_clean = treatments_clean.drop('dose', axis=1)

##### Test

In [161]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  dose_start dose_end  
0        41u      48u  
3        33u      36u  
6        37u      42u  
7        31u      38u  
9        30u      36u

#### Adverse reaction should be part of the `treatments` table

##### Define
Merge the *adverse_reaction* column to the `treatments` table, joining on *given name* and *surname*.

##### Code

In [162]:
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean,
                            on=['given_name', 'surname'], how='left')

##### Test

In [163]:
treatments_clean

given_name        surname  hba1c_start  hba1c_end  hba1c_change  \
0        veronika       jindrová         7.63       7.20          0.43   
1            skye     gormanston         7.97       7.62          0.35   
2          sophia         haugen         7.65       7.27          0.38   
3           eddie         archer         7.89       7.55          0.34   
4            asia        woźniak         7.76       7.37          0.39   
5          joseph            day         7.70       7.19          0.51   
6         roxanne      andreyeva         9.54       9.14          0.40   
7          simone   baumgaertner         7.74       7.30          0.44   
8            enco         žibrik         7.78       7.34          0.44   
9         camilla       zaitseva         7.53       7.13          0.40   
10          tekla        walczak         7.61       7.29          0.32   
11    brancaleone          russo         8.61       8.18          0.43   
12           isac           berg         9.68       9.29          0.39   
13        clinton         miller         7.79       7.40          0.39   
14         eugene        mironov         7.81       7.48          0.33   
15    szilveszter          totth         7.70       7.38          0.32   
16      alexander      mathiesen         7.96       7.55          0.41   
17            chỉ            lâm         7.68       7.24          0.44   
18      władysław      wieczorek         7.92       7.47          0.45   
19       kristján        ingason         7.92       7.57          0.35   
20         marija       grubišić         7.53       7.15          0.38   
21          sauli     koivuniemi         7.67       7.37          0.30   
22        mariana          souza         7.86       7.51          0.35   
23     kristoffer      martinsen         9.18       8.64          0.54   
24             mỹ          quynh         7.61       7.16          0.45   
25           oles        zhdanov         7.52       7.11          0.41   
26        triana.       terrazas         7.71       7.34          0.37   
27         gabryŝ    tomaszewski         7.87       7.47          0.40   
28      leixandre         alanis         7.74       7.32          0.42   
29   onyekachukwu         obinna         7.58       7.12          0.46   
..            ...            ...          ...        ...           ...   
320          jane        citizen         7.98       7.60          0.38   
321        angela     lavrentyev         7.61       7.14          0.47   
322        edelma    villalpando         7.99       7.56          0.43   
323        annika          vaara         7.73       7.34          0.39   
324         chiho           higa         7.71       7.30          0.41   
325     beatrycze        woźniak         7.54       7.17          0.37   
326      miłosław     wiśniewski         7.51       7.08          0.43   
327       firenze          fodor         7.89       7.55          0.34   
328           zoe        wellish         7.71       7.30          0.41   
329           una  traustadóttir         8.00       7.50          0.50   
330         luboš          pecha         7.79       7.45          0.34   
331         meaza         brhane         7.70       7.36          0.34   
332         adlan       shishani         7.84       7.37          0.47   
333         sofia      hermansen         8.90       8.57          0.33   
334         guðni      heimisson         7.64       7.24          0.40   
335       eufemio        rosario         7.54       7.26          0.28   
336      dalmacia         madrid         7.67       7.21          0.46   
337         daimy          tromp         9.41       8.94          0.47   
338        jeremy        montagu         7.68       7.36          0.32   
339       nebechi      ekechukwu         7.78       7.39          0.39   
340       satsita      batukayev         7.63       7.25          0.38   
341       timothy         cotton         7.92       7.52          0.40   
342       b

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
Adverse reactions table is no longer needed so ignore that part. Isolate the patient ID and names in the `patients` table, then convert these names to lower case to join with `treatments`. Then drop the given name and surname columns in the treatments table (so these being lowercase isn't an issue anymore).

##### Code

In [164]:
id_names = patients_clean[['patient_id', 'given_name', 'surname']]
id_names.given_name = id_names.given_name.str.lower()
id_names.surname = id_names.surname.str.lower()
treatments_clean = pd.merge(treatments_clean, id_names, on=['given_name', 'surname'])
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)

/opt/conda/lib/python3.6/site-packages/pandas/core/generic.py:4405: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


##### Test

In [165]:
# Confirm the merge was executed correctly
treatments_clean

hba1c_start  hba1c_end  hba1c_change treatment dose_start dose_end  \
0           7.63       7.20          0.43   auralin        41u      48u   
1           7.97       7.62          0.35   auralin        33u      36u   
2           7.65       7.27          0.38   auralin        37u      42u   
3           7.89       7.55          0.34   auralin        31u      38u   
4           7.76       7.37          0.39   auralin        30u      36u   
5           7.70       7.19          0.51   auralin        29u      36u   
6           7.70       7.19          0.51   auralin        29u      36u   
7           9.54       9.14          0.40   auralin        29u      38u   
8           7.74       7.30          0.44   auralin        27u      37u   
9           7.78       7.34          0.44   auralin        55u      68u   
10          7.53       7.13          0.40   auralin        28u      37u   
11          7.61       7.29          0.32   auralin        29u      39u   
12          8.61       8.18          0.43   auralin        53u      60u   
13          9.68       9.29          0.39   auralin        31u      41u   
14          7.79       7.40          0.39   auralin        42u      51u   
15          7.81       7.48          0.33   auralin        42u      49u   
16          7.70       7.38          0.32   auralin        35u      39u   
17          7.96       7.55          0.41   auralin        47u      58u   
18          7.68       7.24          0.44   auralin        45u      48u   
19          7.92       7.47          0.45   auralin        24u      37u   
20          7.92       7.57          0.35   auralin        44u      55u   
21          7.53       7.15          0.38   auralin        37u      43u   
22          7.67       7.37          0.30   auralin        43u      47u   
23          7.86       7.51          0.35   auralin        36u      42u   
24          9.18       8.64          0.54   auralin        29u      37u   
25          7.61       7.16          0.45   auralin        57u      64u   
26          7.52       7.11          0.41   auralin        54u      67u   
27          7.71       7.34          0.37   auralin        34u      42u   
28          7.87       7.47          0.40   auralin        29u      37u   
29          7.74       7.32          0.42   auralin        61u      67u   
..           ...        ...           ...       ...        ...      ...   
319         7.98       7.60          0.38   novodra        37u      38u   
320         7.61       7.14          0.47   novodra        28u      24u   
321         7.99       7.56          0.43   novodra        24u      26u   
322         7.73       7.34          0.39   novodra        20u      21u   
323         7.71       7.30          0.41   novodra        46u      46u   
324         7.54       7.17          0.37   novodra        26u      27u   
325         7.51       7.08          0.43   novodra        34u      33u   
326         7.89       7.55          0.34   novodra        30u      35u   
327         7.71       7.30          0.41   novodra        33u      33u   
328         8.00       7.50          0.50   novodra        35u      34u   
329         7.79       7.45          0.34   novodra        30u      27u   
330         7.70       7.36          0.34   novodra        37u      41u   
331         7.84       7.37          0.47   novodra        43u      40u   
332         8.90       8.57          0.33   novodra        34u      34u   
333         7.64       7.24          0.40   novodra        40u      36u   
334         7.54       7.26          0.28   novodra        37u      40u   
335         7.67       7.21          0.46   novodra        26u      23u   
336         9.41       8.94          0.47   novodra        40u      45u   
337         7.68       7.36          0.32   novodra        52u      52u   
338         7.78       7.39          0.39   novodra        37u      39u   
339         7.63       7.25          0.38   novodra        42u      42u   
340         7.92       7.52   

In [166]:
# Patient ID should be the only duplicate column
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

22    patient_id
dtype: object

### Quality

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
Convert the zip code column's data type from a float to a string using `astype`, remove the '.0' using string slicing, and pad four digit zip codes with a leading 0.

In [167]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  birthdate  weight  \
0  Rancho California  California   92390.0  United States  7/10/1976   121.7   
1          Armstrong    Illinois   61812.0  United States   4/3/1967   118.8   
2               York    Nebraska   68467.0  United States  2/19/1980   177.8   
3         Woodbridge          NJ    7095.0  United States  7/26/1951   220.9   
4             Dothan          AL   36303.0  United States  2/18/1928   192.3   

   height   bmi       phone_number                      email  
0      66  19.6       951-719-9170   ZoeWellish@superrito.com  
1      66  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2      71  24.8       402-363-6804       JaeMDebord@gustr.com  
3      70  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4      27  26.1       334-515-7487        TimNeudorf@cuvox.de

##### Code

In [168]:
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5, fillchar='0')
# Reconvert NaNs entries that were converted to '0000n' by code above
patients_clean.zip_code = patients_clean.zip_code.replace('0000n', np.nan)

##### Test

In [169]:
patients_clean.zip_code.head()

0    92390
1    61812
2    68467
3    07095
4    36303
Name: zip_code, dtype: object

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
Replace height for rows in the `patients` table that have a height of 27 in (there is only one) with 72 in.

##### Code

In [170]:
patients_clean.height = patients_clean.height.replace(27, 72)

##### Test

In [171]:
# Should be empty
patients_clean[patients_clean.height == 27]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [172]:
# Confirm the replacement worked
patients_clean[patients_clean.surname == 'Neudorf']

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      72  26.1   

   phone_number                email  
4  334-515-7487  TimNeudorf@cuvox.de

#### Full state names sometimes, abbreviations other times

##### Define
Apply a function that converts full state name to state abbreviation for California, New York, Illinois, Florida, and Nebraska.

##### Code

In [173]:
# Mapping from full state name to abbreviation
state_abbrev = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}

# Function to apply
def abbreviate_state(patient):
    if patient['state'] in state_abbrev.keys():
        abbrev = state_abbrev[patient['state']]
        return abbrev
    else:
        return patient['state']
    
patients_clean['state'] = patients_clean.apply(abbreviate_state, axis=1)

##### Test

In [174]:
patients_clean.state.value_counts()

CA    60
NY    47
TX    32
IL    24
MA    22
FL    22
PA    18
GA    15
OH    14
LA    13
OK    13
MI    13
NJ    12
VA    11
WI    10
MS    10
MN     9
TN     9
IN     9
AL     9
KY     8
WA     8
NC     8
MO     7
NE     6
KS     6
NV     6
ID     6
CT     5
SC     5
IA     5
AR     4
CO     4
ND     4
ME     4
AZ     4
RI     4
MD     3
SD     3
DE     3
WV     3
OR     3
MT     2
VT     2
DC     2
AK     1
WY     1
NM     1
NH     1
Name: state, dtype: int64

#### Dsvid Gustafsson

##### Define
Replace given name for rows in the `patients` table that have a given name of 'Dsvid' with 'David'.

##### Code

In [175]:
patients_clean.given_name = patients_clean.given_name.replace('Dsvid', 'David')

##### Test

In [176]:
patients_clean[patients_clean.surname == 'Gustafsson']

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

   phone_number                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
Convert assigned sex and state to categorical data types. Zip code data type was already addressed above. Convert birthdate to datetime data type. Strip the letter 'u' in start dose and end dose and convert those columns to data type integer.

##### Code

In [177]:
# To category
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')

# To datetime
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

# Strip u and to integer
treatments_clean.dose_start = treatments_clean.dose_start.str.strip('u').astype(int)
treatments_clean.dose_end = treatments_clean.dose_end.str.strip('u').astype(int)

In [178]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city state zip_code        country  birthdate  weight  height  \
0  Rancho California    CA    92390  United States 1976-07-10   121.7      66   
1          Armstrong    IL    61812  United States 1967-04-03   118.8      66   
2               York    NE    68467  United States 1980-02-19   177.8      71   
3         Woodbridge    NJ    07095  United States 1951-07-26   220.9      70   
4             Dothan    AL    36303  United States 1928-02-18   192.3      72   

    bmi       phone_number                      email  
0  19.6       951-719-9170   ZoeWellish@superrito.com  
1  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2  24.8       402-363-6804       JaeMDebord@gustr.com  
3  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4  26.1       334-515-7487        TimNeudorf@cuvox.de

In [179]:
treatments_clean.head()

hba1c_start  hba1c_end  hba1c_change treatment  dose_start  dose_end  \
0         7.63       7.20          0.43   auralin          41        48   
1         7.97       7.62          0.35   auralin          33        36   
2         7.65       7.27          0.38   auralin          37        42   
3         7.89       7.55          0.34   auralin          31        38   
4         7.76       7.37          0.39   auralin          30        36   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242  
2              NaN         345  
3              NaN         276  
4              NaN          15

##### Test

In [180]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null category
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null category
zip_code        491 non-null object
country         491 non-null object
birthdate       503 non-null datetime64[ns]
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phone_number    491 non-null object
email           491 non-null object
dtypes: category(2), datetime64[ns](1), float64(2), int64(2), object(8)
memory usage: 53.9+ KB


In [181]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
hba1c_start         349 non-null float64
hba1c_end           349 non-null float64
hba1c_change        349 non-null float64
treatment           349 non-null object
dose_start          349 non-null int64
dose_end            349 non-null int64
adverse_reaction    35 non-null object
patient_id          349 non-null int64
dtypes: float64(3), int64(3), object(2)
memory usage: 24.5+ KB


#### Multiple phone number formats

##### Define
Strip all " ", "-", "(", ")", and "+" and store each number without any formatting. Pad the phone number with a 1 if the length of the number is 10 digits (we want country code).

##### Code

In [182]:
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '').str.pad(11, fillchar='1')

##### Test

In [183]:
patients_clean.phone_number.head()

0    19517199170
1    12175693204
2    14023636804
3    17326368246
4    13345157487
Name: phone_number, dtype: object

#### Default John Doe data

##### Define
Remove the non-recoverable John Doe records from the `patients` table.

##### Code

In [184]:
patients_clean = patients_clean[patients_clean.surname != 'Doe']

##### Test

In [185]:
# Should be no Doe records
patients_clean.surname.value_counts()

Jakobsen        3
Taylor          3
Johnson         2
Kadyrov         2
Tạ              2
Gersten         2
Bùi             2
Cabrera         2
Hueber          2
Dratchev        2
Tucker          2
Liễu            2
Lương           2
Ogochukwu       2
Lund            2
Grímsdóttir     2
Silva           2
Aranda          2
Souza           2
Berg            2
Schiavone       2
Nilsen          2
Collins         2
Kowalczyk       2
Batukayev       2
Correia         2
Parker          2
Lâm             2
Woźniak         2
Cindrić         2
               ..
Shishani        1
Tuma            1
Achebe          1
Nielsen         1
Brhane          1
Pospíšil        1
Nebay           1
Fodor           1
Csonka          1
Azevedo         1
Bole            1
Okoli           1
Crawford        1
Kristiansen     1
Ferreira        1
Grant           1
Chandler        1
Tansey          1
Baum            1
O'Brian         1
Onyekaozulu     1
Sleiman         1
Ekechukwu       1
Tobeolisa       1
Koldenhof 

In [186]:
# Should be no 123 Main Street records
patients_clean.address.value_counts()

2778 North Avenue           2
2476 Fulton Street          2
648 Old Dear Lane           2
1094 Jones Avenue           1
1324 Duffy Street           1
2808 Circle Drive           1
2166 Tipple Road            1
4982 Wood Street            1
1763 Williams Mine Road     1
631 Isaacs Creek Road       1
163 Hide A Way Road         1
4851 Andy Street            1
4970 Heather Sees Way       1
4435 Poe Road               1
200 Hall Place              1
2881 Harter Street          1
3683 Gorby Lane             1
192 Patton Lane             1
57 Norman Street            1
846 Copperhead Road         1
4890 Abia Martin Drive      1
3077 Colony Street          1
2784 Turkey Pen Road        1
522 Lamberts Branch Road    1
4500 Myra Street            1
3261 Desert Broom Court     1
288 Colony Street           1
3259 Roy Alley              1
541 Green Hill Road         1
1813 Lindale Avenue         1
                           ..
3006 Maple Court            1
230 Lady Bug Drive          1
3094 Oral 

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
Remove the Jake Jakobsen, Pat Gersten, and Sandy Taylor rows from the `patients` table. These are the nicknames, which happen to also not be in the `treatments` table (removing the wrong name would create a consistency issue between the `patients` and `treatments` table). These are all the second occurrence of the duplicate. These are also the only occurences of non-null duplicate addresses.

##### Code

In [187]:
# tilde means not: http://pandas.pydata.org/pandas-docs/stable/indexing.html#boolean-indexing
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

##### Test

In [188]:
patients_clean[patients_clean.surname == 'Jakobsen']

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
432      Houston    TX    77020  United States 1962-11-25   185.2      67   

      bmi phone_number                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

In [189]:
patients_clean[patients_clean.surname == 'Gersten']

patient_id assigned_sex given_name  surname            address  city  \
97          98         male    Patrick  Gersten  2778 North Avenue  Burr   

   state zip_code        country  birthdate  weight  height   bmi  \
97    NE    68324  United States 1954-05-03   138.2      71  19.3   

   phone_number                     email  
97  14028484923  PatrickGersten@rhyta.com

In [190]:
patients_clean[patients_clean.surname == 'Taylor']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

    state zip_code        country  birthdate  weight  height   bmi  \
131    WV    25962  United States 1960-10-23   206.1      64  35.4   
426    FL    33179  United States 1992-09-02   186.6      69  27.6   

    phone_number                       email  
131  13044382648    SandraCTaylor@dayrep.com  
426  13054346299  RogelioJTaylor@teleworm.us

#### kgs instead of lbs for Zaitseva weight

##### Define
Use [advanced indexing](https://stackoverflow.com/a/44913631) to isolate the row where the surname is Zaitseva and convert the entry in its weight field from kg to lbs.

##### Code

In [191]:
#weight_kg = patients_clean.weight.min()
weight_kg = patients_clean.weight.min()
mask = patients_clean.surname == 'Zaitseva'
column_name = 'weight'
patients_clean.loc[mask, column_name] = weight_kg * 2.20462

In [192]:
patients_clean[patients_clean.surname == 'Zaitseva']

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country  birthdate      weight  height  \
210  Wooster    OH    44691  United States 1938-11-26  107.585456      63   

      bmi phone_number                          email  
210  19.1  13302022145  CamillaZaitseva@superrito.com

##### Test

In [193]:
# 48.8 shouldn't be the lowest anymore
patients_clean.weight.sort_values()

459    102.100000
335    102.700000
74     103.200000
317    106.000000
171    106.500000
51     107.100000
210    107.585456
270    108.100000
198    108.500000
48     109.100000
478    109.600000
141    110.200000
38     111.800000
438    112.000000
14     112.000000
235    112.200000
307    112.400000
191    112.600000
408    113.100000
49     113.300000
326    114.000000
338    114.100000
253    117.000000
321    118.400000
168    118.800000
1      118.800000
350    119.000000
207    119.200000
265    120.000000
341    120.300000
          ...    
332    224.000000
12     224.200000
252    224.200000
222    224.800000
166    225.300000
111    225.900000
101    226.200000
150    226.600000
88     227.700000
352    227.700000
428    227.700000
13     228.400000
339    229.000000
182    230.300000
121    230.800000
257    231.700000
395    231.900000
246    232.100000
219    237.800000
11     238.700000
50     238.900000
441    239.100000
499    239.600000
439    242.000000
487    242